cibyl/models/ci/system.py

```diff
@@ -13,180 +13,197 @@
 #    License for the specific language governing permissions and limitations
 #    under the License.
 """
-# pylint: disable=no-member
-from abc import abstractmethod
-from typing import Dict, List
+from abc import abstractmethod, ABC
+from typing import Dict, List, Type
 
 from cibyl.cli.argument import Argument
 from cibyl.exceptions.model import NonSupportedModelType
 from cibyl.models.attribute import AttributeDictValue, AttributeListValue
 from cibyl.models.ci.job import Job
-from cibyl.models.ci.pipeline import Pipeline
 from cibyl.models.model import Model
 from cibyl.sources.source import Source
-from cibyl.utils.colors import Colors
+
+BASE_SYSTEM_API = {
+    'name': {
+        'attr_type': str,
+        'arguments': []
+    },
+    'system_type': {
+        'attr_type': str,
+        'arguments': [
+            Argument(name='--system-type', arg_type=str,
+                     description="System type")
+        ]
+    },
+    'sources': {
+        'attr_type': Source,
+        'attribute_value_class': AttributeListValue,
+        'arguments': [
+            Argument(name='--sources', arg_type=str,
+                     nargs="*",
+                     description="Source name")
+        ]
+    }
+}
+
+JOBS_SYSTEM_API = {
+    'jobs_scope': {
+        'attr_type': str,
+        'arguments': []
+    },
+    'jobs': {
+        'attr_type': Job,
+        'attribute_value_class': AttributeDictValue,
+        'arguments': [
+            Argument(name='--jobs', arg_type=str,
+                     nargs='*',
+                     description="System jobs",
+                     func='get_jobs')
+        ]
+    }
+}
+
+ZUUL_SYSTEM_API = {
+}
 
 
 class System(Model):
-    """General model for a CI system.
+    """Base model for a CI system.
 
-    Holds basic information such as its name, type and which jobs it has.
+    Holds basic information such as the system's name or its type.
+
+    :ivar top_level_model: Defines the model that is the direct child of
+        this system. For most system's, this will be :class:`Job`.
     """
     API = {
```

Comment:
If we include everything in the System API, we will always have all the arguments shown. That is, even if there is no zuul system in the user configuration we will still show the --tenant, or whatever specific arguments a Zuul system has

Reply:
Yes, that is a trade off that I had to accept. It is something that I expended time trying to avoid but there is no easy way to. The thing is that the Environment API can only hold one child API for the '--systems' argument. The one chosen right now is the API from System. I have tried different ways of solving this, but they all break everything above the Environment class and I do not want to deal with that at this point.

It is not a huge bug, the only problem is that arguments for unwanted systems will be printed. We can report it and study it later on.

Reply:
The pipeline model will surely appear again at some other point, but not with the same meaning as it has right now. What we called Pipeline before is now called a Tenant.

Reply:
That issue with the API can be solved by modifying the base System API in the constructor of the Zuul system and having  the JobsSystem inherit directly the base System API without the Zuul-specific arguments as it was before. In this way, the arguments for jenkins systems will always be shown and the zuul-specific ones will appear if there is a zuul system in use.

Reply:
Thing is that the CLI parser never visits either the JobSystem or the ZuulSystem classes. Everything has to be set up on the API variable at the System class on one go. I cannot use constructors here. The problem was still there before my changes, the previous solution was to have all parameters defined on the System class and the let the others pick which ones they wanted. Only thing I did was to move those arguments to their relative classes so that it is easier to keep track of.

Reply:
You're right that the parser only uses the base System, but the ZuulSystem constructor could be used to modify the base System API, which is what was done before. The environments and systems are constructed before the parser is extended, so the constructor can be used for that.

Reply:
On any case, tests pass on the way I have implemented it and as far as I can see the previous implementation did not fix the problem we have with the CLI. So I am going to keep this for the time being unless it can become a problem.

Reply:
But the previous implementation did solve the problem, it would show just jobs or the top-level model for zuul (in that case it was pipeline, now it's tenant) if a zuul system was used.